src/meta/src/rpc/server.rs

```diff
@@ -171,316 +153,312 @@ pub async fn rpc_serve_with_store<S: MetaStore>(
+        // wait until this node becomes a leader
+        while !node_is_leader(&leader_rx) {
+            leader_rx.changed().await.expect("Leader sender dropped");
+        }
 
-        let join_handle = tokio::spawn(async move {
-            tokio::select! {
-                _ = tokio::signal::ctrl_c() => {}
-                _ = &mut svc_shutdown_rx => {
-                    shutdown_election.await;
-                }
+        // shut down follower svc if node used to be follower
+        if let Some(handle) = follower_handle {
+            match follower_shutdown_tx.send(()) {
+                Ok(_) => tracing::info!("Shutting down follower services"),
+                Err(_) => tracing::error!("Follower service receiver dropped"),
             }
-        });
+            // Wait until follower service is down
+            handle.await.unwrap();
+        }
 
-        // FIXME: Avoid using join_handler and just pass around shutdown_sender
-        // https://github.com/risingwavelabs/risingwave/pull/6771
-        return Ok((join_handle, svc_shutdown_tx));
-    }
+        let elect_coord = ElectionCoordination {
+            election_handle,
+            election_shutdown,
+        };
 
-    tracing::info!("Node initially elected as leader");
+        let current_leader = services_leader_rx.borrow().0.clone();
+        start_leader_srv(
+            meta_store,
+            address_info,
+            max_heartbeat_interval,
+            opts,
+            current_leader,
+            elect_coord,
+            svc_shutdown_rx,
+        )
+        .await
+        .expect("Unable to start leader services");
+    });
 
-    let env = MetaSrvEnv::<S>::new(opts, meta_store.clone(), current_leader_info).await;
-    let fragment_manager = Arc::new(FragmentManager::new(env.clone()).await.unwrap());
-    let meta_metrics = Arc::new(MetaMetrics::new());
-    let registry = meta_metrics.registry();
-    monitor_process(registry).unwrap();
+    Ok((join_handle, svc_shutdown_tx))
+}
 
-    let cluster_manager = Arc::new(
-        ClusterManager::new(env.clone(), max_heartbeat_interval)
-            .await
-            .unwrap(),
-    );
-    let heartbeat_srv = HeartbeatServiceImpl::new(cluster_manager.clone());
+mod tests {
+    use core::panic;
+    use std::net::{IpAddr, Ipv4Addr};
+
+    use risingwave_common::config::MAX_CONNECTION_WINDOW_SIZE;
+    use risingwave_common::util::addr::HostAddr;
+    use risingwave_pb::common::{HostAddress, WorkerType};
+    use risingwave_pb::meta::cluster_service_client::ClusterServiceClient;
+    use risingwave_pb::meta::AddWorkerNodeRequest;
+    use tokio::time::sleep;
+    use tonic::transport::Endpoint;
+
+    use super::*;
+
+    static _SLEEP_SEC: u64 = 6;
+
+    /// Start `n` meta nodes on localhost. First node will be started at `meta_port`, 2nd node on
+    /// `meta_port + 1`, ...
+    async fn _setup_n_nodes(n: u16, meta_port: u16) -> Vec<(JoinHandle<()>, WatchSender<()>)> {
```

Comment:
```suggestion
    async fn setup_n_nodes(n: u16, meta_port: u16) -> Vec<(JoinHandle<()>, WatchSender<()>)> {
```

Reply:
`_xxxx` is only used for non-used functions or variables in order to pass the rust unused check.

Reply:
I know. That is exactly why I used it 😓 

```
   --> src/meta/src/rpc/server.rs:246:14
    |
246 |     async fn setup_n_nodes(n: u16, meta_port: u16) -> Vec<(JoinHandle<()>, WatchSender<()>)> {
    |              ^^^^^^^^^^^^^
    |
    = note: `#[warn(dead_code)]` on by default
```

Reply:
I will use `#[cfg(test)]` to avoid this issue.